content/oci/oci.go

```diff
@@ -375,6 +376,80 @@ func (s *Store) writeIndexFile() error {
 	return os.WriteFile(s.indexPath, indexJSON, 0666)
 }
 
+// Referrers lists the descriptors of image or artifact manifests directly
+// referencing the given manifest descriptor.
+//
+// fn is called on the referrer results. If artifactType is not empty, only
+// referrers of the same artifact type are fed to fn.
+//
+// Reference: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#listing-referrers
+func (s *Store) Referrers(ctx context.Context, desc ocispec.Descriptor, artifactType string, fn func(referrers []ocispec.Descriptor) error) error {
+	if !descriptor.IsManifest(desc) {
+		return fmt.Errorf("the descriptor %v is not a manifest", desc)
+	}
+	var results []ocispec.Descriptor
+	predecessors, err := s.Predecessors(ctx, desc)
+	if err != nil {
+		return err
+	}
+	for _, node := range predecessors {
+		switch node.MediaType {
```

Comment:
In my implementation, I included a check for the ArtifactType in the descriptor prior to this switch statement as an optimization. Thoughts?

My reasoning was from reading "listing-referrers" from [1.1.0-rc3](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#listing-referrers), "Upon success, the response MUST be a JSON body with an image index containing a list of descriptors. Each descriptor is of an image manifest in the same <name> namespace with a subject field that specifies the value of <digest>. **The descriptors MUST include an artifactType field that is set to the value of the artifactType in the image manifest, if present.** If the artifactType is empty or missing in the image manifest, the value of artifactType MUST be set to the config descriptor mediaType value."

Reply:
In the current implementation, we check the actual manifest content (the json) for the value of `artifactType`. This gives the most accurate value, as the descriptor value of `artifactType` may be missing. Given that the actual manifest is checked, I don't think the descriptor value check is necessary.

Reply:
Correct it is not necessary, but it is aimed to be an optimization that I believe is worthwhile. 

The quote I included from the distribution spec outlines exactly what you have said, "the `artifactType` *may* be missing". As such I believe it is best to only fetch the manifest if it is necessary to do so, i.e. in the case that the `artifactType` *is not missing* it is unnecessary to fetch the manifest as the `artifactType` can be pulled straight from the descriptor. As you already have, in the case that the `artifactType` *is missing* I agree the manifest must be fetched and the `artifactType` be pulled from there.

I believe the only case where it is *necessary* to fetch the manifest each and every time is if the distribution spec allows for the return of a mismatched, non-empty, `artifactType` in the descriptor; which I do not believe this is the case as this would be a major flaw.

Taking another look at my proposed addition, I believe it should be closer to the `artifactType` check at the bottom of the function. In the case that the `artifactType` we use as a filter is empty then no manifests will be fetched and all will be included in the results.

```go
		if artifactType == "" || node.ArtifactType == artifactType {
			results = append(results, node)
			continue
		}
```

Reply:
We have to fetch the manifest each and every time, as we are using `Predecessors()` to implement `Referrers()`. `Predecessors()` returns not only the referrers of the node but also other predecessors. We have to check the content of each manifest, and mark it as a referrer only when the `Subject` field matches. So even if the `ArtifactType` field of the descriptor is present and correct, we have to check the manifest content for its `Subject`.